src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -3123,6 +3125,20 @@ internal static bool CheckValEscape(SyntaxNode node, BoundExpression expr, uint
 
                     return true;
 
+                case BoundKind.WithExpression:
+                    {
+                        var withExpr = (BoundWithExpression)expr;
+                        var escape = CheckValEscape(node, withExpr.Receiver, escapeFrom, escapeTo, checkingReceiver: false, diagnostics);
```

Comment:
>CheckValEscape(node, withExpr.Receiver, escapeFrom, escapeTo, checkingReceiver: false, diagnostics); [](start = 37, length = 100)

Were we missing this check for 'with' on records?

Reply:
From what I understand, val escape analyses are initiated when we do an assignment (we call `GetValEscape` and `ValidateEscape` from `BindAssignment`, `BindDeconstructionAssignment`, `BindReturn`, etc). But it's only relevant if we're assigning a ref-like type (`GetValEscape` returns `ExternalScope` for all other types).
So I think record structs introduce the need to analyze `with` expressions, because the type of the receiver and the `with` expression now can be a ref-like struct.

For the initializer part, I'm not sure. I've added a test `WithExpr_ValEscape` which would have thought should produce diagnostics because the initializer contains dangerous assignments. Could you take a look?
If I'm correct, this is an existing bug (an assignment for which we are not triggering val-escape checks).

---
In reply to: [607968314](https://github.com/dotnet/roslyn/pull/52319#discussion_r607968314) [](ancestors = 607968314)

Reply:
Chatted with Jared. The scenario in `WithExpr_ValEscape` with assigning a ref struct value to a property is fine. The setter is essentially `void Set(S1)`, which cannot do anything dangerous (it could not store `S1` and it returns void so may not sneak it out).

---
In reply to: [608258240](https://github.com/dotnet/roslyn/pull/52319#discussion_r608258240) [](ancestors = 608258240,607968314)